File: admin/app/Http/Controllers/Admin/AdminController.php
<?php

namespace App\Http\Controllers\Admin;

use App\Http\Controllers\Controller;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\Log;

class AdminController extends Controller
{
    private $apiUrl = 'http://172.27.1.162:8080/admin';
    private $requestTable = 'http://172.27.1.162:8080/requestTable';

    public function adminLogin(Request $request)
    {
        if ($request->isMethod('post')) {
            $username = $request->input('username');
            $password = $request->input('password');
            if (empty($username) || empty($password)) {
                return redirect()->back()->with('error', 'Username and password are required.');
            }

            try {
                $response = Http::post("{$this->apiUrl}/login", [
                    'username' => $username,
                    'password' => $password,
                ]);

                if ($response->failed()) {
                    // Log the response to get more details
                    Log::error('Login API failed', ['response' => $response->body()]);
                    return redirect()->back()->with('error', 'Invalid credentials or server error.');
                }

                $data = $response->json();
                session(['jwt' => $data['token']]);
                return redirect('/admin/dashboard')->with('success', 'Login successfully.');
            } catch (\Throwable $th) {
                // Log the exception to get more details
                Log::error('Login failed', ['exception' => $th]);
                return redirect()->back()->with('error', 'Internal Server Error.');
            }
        }

        return view('admin.login');
    }

    public function getProfile(Request $request)
    {
        try {
            $token = session('jwt');
            if (!$token) {
                return redirect('/admin/login')->with('error', 'Unauthenticated');
            }

            $response = Http::withHeaders([
                'Cookie' => "jwt={$token}",
            ])->get("{$this->apiUrl}/profile");

            if ($response->failed()) {
                throw new \Exception("Failed to fetch profile.");
            }

            $data = $response->json();
            return view('admin.account.profile', compact('data'));
        } catch (\Throwable $th) {
            return redirect('/admin/login')->with('error', 'You must be logged in.');
        }
    }

    public function logoutAdmin(Request $request)
    {
        try {
            $token = session('jwt');
            if (!$token) {
                return redirect('/admin/login')->with('error', 'Unauthenticated');
            }

            $response = Http::withHeaders([
                'Cookie' => "jwt={$token}",
            ])->post("{$this->apiUrl}/logout");

            session()->forget('jwt');
            return redirect('/admin/login')->with('success', 'Logout successful');
        } catch (\Throwable $th) {
            return redirect('/admin/login')->with('error', 'You must be logged in.');
        }
    }


    public function profile(Request $request)
    {
        try {
            $token = session('jwt');
            if (!$token) {
                return redirect('/admin/login')->with('error', 'Unauthenticated');
            }

            $response = Http::withHeaders([
                'Authorization' => "Bearer {$token}",
            ])->get("{$this->apiUrl}/profile");

            if ($response->failed()) {
                throw new \Exception("Failed to fetch profile.");
            }

            $data = $response->json();
            return view('admin.account.profile', compact('data'));
        } catch (\Throwable $th) {
            return redirect('/admin/login')->with('error', 'You must be logged in.');
        }
    }

    // Add the dashboard method here

    public function dashboard(Request $request)
    {
        try {
            $token = session('jwt');
            if (!$token) {
                return redirect('/admin/login')->with('error', 'Unauthenticated');
            }
            $response = Http::withHeaders([
                'Cookie' => "jwt={$token}",
            ])->get("{$this->apiUrl}/profile");

            if ($response->failed()) {
                throw new \Exception("Failed to fetch profile.");
            }

            $data = $response->json();

            return view('admin.dashboard', compact('data')); // Ensure you have a 'dashboard.blade.php' file in the 'admin' view directory
        } catch (\Throwable $th) {
            return redirect('/admin/login')->with('error', 'You must be logged in.');
        }
    }

    public function approveReqTable(Request $request, $id)
    {
        try {
            $token = session('jwt');
            if (!$token) {
                return redirect('/admin/login')->with('error', 'Unauthenticated');
            }

            // Convert id to integer
            $id = intval($id);
            if ($id <= 0) {
                return redirect()->back()->with('error', 'Invalid table ID.');
            }

            // Parse input from request
            $input = $request->all();

            // Call the API to approve the request table
            $response = Http::withHeaders([
                'Cookie' => "jwt {$token}",
            ])->put("{$this->apiUrl}/table/{$id}", $input);

            if ($response->failed()) {
                return redirect()->back()->with('error', 'Failed to approve request table. ' . $response->body());
            }

            return redirect()->back()->with('success', 'Request table approved successfully.');
        } catch (\Throwable $th) {
            Log::error('Failed to approve request table', ['exception' => $th]);
            return redirect()->back()->with('error', 'Failed to approve request table.');
        }
    }

    public function getAllRequestTables(Request $request)
    {
        try {
            $token = session('jwt');
            if (!$token) {
                return redirect('/admin/login')->with('error', 'Unauthenticated');
            }

            // Request for admin data
            $adminResponse = Http::withHeaders([
                'Cookie' => "jwt={$token}",
            ])->get("{$this->apiUrl}/profile");

            if ($adminResponse->failed()) {
                dd('Failed to fetch admin data', $adminResponse->body()); // Debugging line
                Log::error('Failed to fetch admin data', ['response' => $adminResponse->body()]);
                return redirect()->back()->with('error', 'Failed to fetch admin data. Please check the logs for more details.');
            }

            // Request for request tables data
            $requestTableResponse = Http::withHeaders([
                'Cookie' => "jwt={$token}",
            ])->get("{$this->requestTable}");

            if ($requestTableResponse->failed()) {
                dd('Failed to fetch request tables', $requestTableResponse->body()); // Debugging line
                Log::error('Failed to fetch request tables', ['response' => $requestTableResponse->body()]);
                return redirect()->back()->with('error', 'Failed to fetch request tables. Please check the logs for more details.');
            }

            $data = $adminResponse->json();
            $requestTables = $requestTableResponse->json();

            // Debugging line to check the response data
            // dd('Admin Data:', $adminData, 'Request Tables:', $requestTables);

            return view('admin.requestTable.index', compact('data', 'requestTables'));
        } catch (\Throwable $th) {
            Log::error('Failed to fetch data', ['exception' => $th]);
            return redirect()->back()->with('error', 'Failed to fetch data.');
        }
    }


    public function approveRequestTable(Request $request, $id)
    {
        return $this->updateRequestTableStatus($request, $id, 1);
    }

    public function rejectRequestTable(Request $request, $id)
    {
        return $this->updateRequestTableStatus($request, $id, 2);
    }

    public function pendingRequestTable(Request $request, $id)
    {
        return $this->updateRequestTableStatus($request, $id, 0);
    }

    private function updateRequestTableStatus(Request $request, $id, $status)
{
    try {
        $token = session('jwt');
        if (!$token) {
            return redirect('/admin/login')->with('error', 'Unauthenticated');
        }

        $id = intval($id);
        if ($id <= 0) {
            return redirect()->back()->with('error', 'Invalid request table ID.');
        }

        Log::info('Updating request table status', ['id' => $id, 'status' => $status]);

        $response = Http::withHeaders([
            'Cookie' => "jwt={$token}",
        ])->put("{$this->apiUrl}/table/{$id}", ['status' => $status]);

        if ($response->failed()) {
            Log::error('Failed to update request table status', ['response' => $response->body()]);
            return redirect()->back()->with('error', 'Failed to update request table status. ' . $response->body());
        }

        return redirect()->back()->with('success', 'Request table status updated successfully.');
    } catch (\Throwable $th) {
        Log::error('Failed to update request table status', ['exception' => $th]);
        return redirect()->back()->with('error', 'Failed to update request table status.');
    }
}

}

File: The Deck/Controllers/AdminController.go
package controllers

import (
	"api/the_deck/Models/dto"
	service "api/the_deck/Service"
	utils "api/the_deck/Utils"
	"strconv"

	"github.com/go-playground/validator/v10"
	"github.com/gofiber/fiber/v2"
)

type AdminController struct {
	adminService service.AdminService
}

func NewAdminController(as service.AdminService) *AdminController {
	return &AdminController{adminService: as}
}

func (c *AdminController) AdminLogin(ctx *fiber.Ctx) error {
	var input dto.RequestAdminLogin
	if err := ctx.BodyParser(&input); err != nil {
		return utils.MessageJSON(ctx, fiber.StatusBadRequest, "Failed", "Invalid request body")
	}

	validation := validator.New()
	if err := validation.Struct(input); err != nil {
		return ctx.Status(400).JSON(fiber.Map{
			"status":  "failed",
			"message": err.Error(),
		})
	}

	admin, token, err := c.adminService.AdminLogin(ctx, input)
	if err != nil {
		return utils.MessageJSON(ctx, fiber.StatusUnauthorized, "Failed", err.Error())
	}

	return ctx.JSON(fiber.Map{
		"status":  "message",
		"message": admin,
		"token":   token,
	})
}

func (c *AdminController) GetProfile(ctx *fiber.Ctx) error {
	admin, err := c.adminService.GetProfile(ctx)
	if err != nil {
		return utils.MessageJSON(ctx, 404, "Failed", "Unauthenticated")
	}

	return ctx.Status(200).JSON(fiber.Map{
		"status":  "success",
		"message": admin,
	})
}

func (c *AdminController) LogoutAdmin(ctx *fiber.Ctx) error {
	cookie, err := c.adminService.LogoutAdmin(ctx)
	if err != nil {
		return err
	}

	ctx.Cookie(cookie)

	ctx.Locals("admin", nil)

	return ctx.Status(200).JSON(fiber.Map{
		"status":  "success",
		"message": "Logout successfully",
	})
}

func (c *AdminController) UpdateStatusOrder(ctx *fiber.Ctx) error {
	id, err := strconv.Atoi(ctx.Params("id"))
	if err != nil {
		return utils.MessageJSON(ctx, 400, "Failed", err.Error())
	}

	var input dto.RequestOrderUpdateStatus
	if err := ctx.BodyParser(&input); err != nil {
		return err
	}

	order, err := c.adminService.UpdateStatus(ctx, uint(id), input)
	if err != nil {
		return utils.MessageJSON(ctx, 400, "Failed", err.Error())
	}

	return ctx.Status(200).JSON(fiber.Map{
		"status":  "success",
		"message": order,
	})
}

func (c *AdminController) ApproveReqTable(ctx *fiber.Ctx) error {
	id, err := strconv.Atoi(ctx.Params("id"))
	if err != nil {
		return err
	}

	var input dto.UpdateRequestTable
	if err := ctx.BodyParser(&input); err != nil {
		return err
	}

	req_table, err := c.adminService.ApproveReqTable(ctx, uint(id), input)
	if err != nil {
		return utils.MessageJSON(ctx, 400, "Failed", err.Error())
	}

	return ctx.Status(200).JSON(fiber.Map{
		"status":  "success",
		"message": req_table,
	})
}
